### PR TITLE
Feat: Re-enabling gravity bridge conditionally

### PR DESCRIPTION
**Pull request by thomas-nguy**
We probably needs to add more test in case gravity is not enable

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Comment by thomas-nguy** [comment]
@tomtau should the ADR be updated as well?

**Comment by tomtau** [comment]
> @tomtau should the ADR be updated as well?

If there's a need for it, yes: the record doesn't say anything about conditional compilation: https://github.com/crypto-org-chain/cronos/blob/main/docs/architecture/adr-001.md#decision

The original draft had the conditional compilation, but from the discussion, it turned out that there was no need for it:

- the current "mainnet beta" functionality can be forked into a "release/v0.6.x" branch and potential non-breaking changes from "main" can be cherry-picked in PRs onto it (e.g. switching to the next 0.7 Ethermint upstream release)
- the latest/testnet development can just continue on main

**Comment by thomas-nguy** [comment]
Per discussion, I thought that we wanted the conditional compilation because we still don't know when gravity will be mature enough to be used in production and there are chances that breaking changes will need to be ported to mainnet before that

If we want to avoid maintaining multiple fork, this would be the easiest solution?

**Comment by tomtau** [comment]
> Per discussion, I thought that we wanted the conditional compilation because we still don't know when gravity will be mature enough to be used in production and there are chances that breaking changes will need to be ported to mainnet before that
> 
> If we want to avoid maintaining multiple fork or doing cherry picking, this would be the easiest solution?

There's a need to maintain a release branch for 0.6.x once "main" runs on the latest Ethermint and has /x/feemarket added. So a release branch is likely expected or unavoidable irrespective of GB.
If we anticipate GB isn't immediately added in a network upgrade altogether with other breaking changes on "main" and we need to keep up the latest development with it (e.g. for testnet binaries)... then yes, conditional feature-guarded compilation makes sense and this clarification should be updated to ADR-001.

File: docs/architecture/adr-001.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## Changelog
 * 12-10-2021: Initial Draft
 * 13-10-2021: Code deletion instead of conditional compilation
+* 13-10-2021: Conditional compilation
```
Comment (tomtau) [inline docs/architecture/adr-001.md:6]:
```suggestion
* 16-11-2021: Conditional compilation
```

File: docs/architecture/adr-001.md [hunk 2]
```diff
@@ -23,7 +26,17 @@ In addition to that, the "x/cronos" module contains Gravity Bridge-related code
 
 Existing integration tests related to Gravity Bridge can be temporarily disabled and later enabled when Gravity Bridge is added back.
 
-Once the Gravity Bridge code is added back, the x/cronos module `ConsensusVersion` should be increased and the corresponding upgrade handler should be added.
+Once the Gravity Bridge code is added back (after mainnet launch), the x/cronos module `ConsensusVersion` should be increased and the corresponding upgrade handler should be added.
```
Comment (tomtau) [inline docs/architecture/adr-001.md:29]: should those (module version + upgrade handler) be conditionally added/changed as well?

Comment (thomas-nguy) [reply docs/architecture/adr-001.md:29]: I think we can avoid this for now. 

We could keep the same consensus version for both binary (with and without gravity) and increase it when we decide to release gravity to mainnet (at that time, we can also remove the conditional compilation)

Comment (tomtau) [reply docs/architecture/adr-001.md:29]: ok, then it's good to move or rephrase this sentence

Comment (thomas-nguy) [reply docs/architecture/adr-001.md]: sure, changed

**Comment by yihuang** [comment]
But I guess the unit tests and integration tests running in CI only target one of the modes, the other mode will not be tested.
For example if we pick the GRAVITY_BRIDGE mode to be the default one, the non gravity bridge mode is not tested in CI.
Unless we run the tests on both modes, will that be overkill?

**Review by yihuang — COMMENTED** [review]
But I guess the unit tests and integration tests running in CI only target one of the modes, the other mode will not be tested.
For example if we pick the gravity-bridge mode to be the default one, the non-gravity-bridge mode won't be tested in CI, and could be silently broken.
Unless we run the tests on both modes, will that be overkill?

**Comment by thomas-nguy** [comment]
Yeah as I mentioned in the PR, we need to open another issue to improve the integration tests.
I guess the easiest way is to add another build + integration tests run in the CI for non gravity bridge binary?

What do you think?

**Comment by yihuang** [comment]
> Yeah as I mentioned in the PR, we need to open another issue to improve the integration tests. I guess the easiest way is to add another build + integration tests run in the CI for non gravity bridge binary?
> 
> What do you think?

Do you think the two modes are temporary or permanent? my concern is it's not easy to keep both modes working correctly in the long run.

**Comment by thomas-nguy** [comment]
`Do you think the two modes are temporary or permanent? my concern is it's not easy to keep both modes working correctly in the long run.`

I guess once we decide to launch main net with gravity bridge, we don't need the conditional build anymore and can just stick with one version

**Comment by yihuang** [comment]
> `Do you think the two modes are temporary or permanent? my concern is it's not easy to keep both modes working correctly in the long run.`
> 
> I guess once we decide to launch main net with gravity bridge, we don't need the conditional build anymore and can just stick with one version

Then I guess using branches should be good enough?
- Before we decide to use the gravity bridge, we develop that in a feature branch.
- After we decide to use the gravity bridge, we create a release branch for non-gravity-bridge version, and merge the gravity bridge version to main branch

**Comment by thomas-nguy** [comment]
Yeah this is possible but the conditional build is just to avoid us to jungle with two branches and cherry-pick back and forth for them to stay sync for a certain amount of time.

But if we decide to go that way, I can remove the conditional build. What do you think @tomtau @calvinaco

**Comment by tomtau** [comment]
if this repo follows TBD, then long-lived feature branches aren't desirable.
if there is a concern for not always building with the feature on, it can be a temporary runtime flag: https://trunkbaseddevelopment.com/feature-flags/  -- and selected E2E or integration tests can be re-executed with that flag on

**Comment by yihuang** [comment]
> if this repo follows TBD, then long-lived feature branches aren't desirable.

It's temporary, it'll merge to the main branch when we decide to use gravity-bridge in the next release.

**Comment by tomtau** [comment]
if temporary feature branch == a few days, then it's all right as TBD: https://trunkbaseddevelopment.com/youre-doing-it-wrong/#duration-of-short-lived-feature-branches

Otherwise, the contribution guidelines may need changes: https://github.com/crypto-org-chain/cronos/blob/main/CONTRIBUTING.md (which branch should one open a PR against if there are two parallel developments? who/when/what back ports changes from one to the other? how are releases made ? ...)

**Comment by yihuang** [comment]
> if temporary feature branch == a few days, then it's all right as TBD: https://trunkbaseddevelopment.com/youre-doing-it-wrong/#duration-of-short-lived-feature-branches
> 
> Otherwise, the contribution guidelines may need changes: https://github.com/crypto-org-chain/cronos/blob/main/CONTRIBUTING.md (which branch should one open a PR against if there are two parallel developments? who/when/what back ports changes from one to the other? how are releases made ? ...)

I guess before we decide to use gravity bridge, there won't be heavy development on it.
I think the blocker of that decision is to evaluate different gravity bridge forks, it seems good to do the evaluation on a feature branch rather than the main branch.

**Review by yihuang — APPROVED** [review]
LGTM

**Pull request merged**